### DISCUSSION TITLE
Add user to a list only if they're not already on it?

### DISCUSSION THREAD

**Discussion by jonathansadowski** [bluesky-social/atproto#3299] [How to do it?] [answered]
Reading over the [documentation for user lists](https://docs.bsky.app/docs/tutorials/user-lists), I've noticed that you can add a subject to a list multiple times. If you've added the subject to the list multiple times, they'll only appear in the UI as on the list once, but if you delete them from the list from the UI, they remain on the list unless you repeat the action for the number of times you've added them.

I also notice that the delete request requires the record key, not simply the actor. Unless I'm missing something, this seems to imply that for both adding somebody to a list, and deleting them, I need to iterate over the entire existing list — in the case of adding, to make sure that I don't add somebody that's already on the list, and in the case of deleting, in order to get the rkey(s) of the subject for deletion.

Is there no way to search the list for the subject without making several API calls to fetch the entire list first? As a list grows, to say, 500 subjects, adding somebody while avoiding duplication would require 6 API calls.

**Comment by DavidBuchanan314** [comment, marked as answer]
there's a trick to avoid this problem - generate the list item rkey as a deterministic function of the subject and the list rkey. you can use `putRecord` for adding users, and it'll do nothing if you add the same subject twice, because it'll have the same rkey.

(however, this only works as long as nobody else's code is adding new entries)

**Reply by jonathansadowski** [reply]
Thanks. In this case I'm controlling all new entries, so I'll give this a shot. It'd be nice if there were a better way to manage things like this, that have an implied uniqueness, but this should suffice for this case.
